### PR TITLE
[Performance] Export Model constructor throws exceptions on very hot paths

### PR DESCRIPTION
**Pull request by i386**
Removes additional millis per call on large exports by avoiding throwing `NotExportableException` and instead using annotation check to check exportability e.g. `type.getAnnotation(ExportedBean.class)`

See https://github.com/jenkinsci/blueocean-plugin/pull/1323

@reviewbybees #NightTimeCode

### CODE REVIEW

File: core/src/main/java/org/kohsuke/stapler/export/Property.java [hunk 1]
```diff
@@ -199,118 +198,127 @@ private void writeValue(Type expected, Object value, TreePruner pruner, DataWrit
...
             }
-            if(value instanceof Iterable) {
-                writer.startArray();
-                Type expectedItemType = Types.getTypeArgument(expected, 0, null);
-                for (Object item : pruner.getRange().apply((Iterable) value)) {
-                    writeBuffered(expectedItemType, item, pruner, writer);
+        }
+
+        try {
+            writer.type(expected, value.getClass());
```
Comment (vivek) [inline core/src/main/java/org/kohsuke/stapler/export/Property.java]: This looks like 🐛 `writer.type()` should be called before writing the object on line 209.

File: core/src/main/java/org/kohsuke/stapler/export/Property.java [hunk 2]
```diff
@@ -199,118 +198,127 @@ private void writeValue(Type expected, Object value, TreePruner pruner, DataWrit
...
-                    for (Object item : r.apply((Object[]) value)) {
-                        writeBuffered(act, item, pruner, writer);
-                    }
-                } else {
-                    // more generic case
-                    int len = Math.min(r.max, Array.getLength(value));
-                    for (int i=r.min; i<len; i++) {
-                        writeBuffered(act, Array.get(value, i), pruner, writer);
-                    }
+            writer.type(expected, value.getClass());
```
Comment (vivek) [inline core/src/main/java/org/kohsuke/stapler/export/Property.java]: It should move down inside else condition below, like 211. handleNotExportable has its own logic to write type.

Comment (i386) [inline core/src/main/java/org/kohsuke/stapler/export/Property.java]: right you are. Moved.

**Comment by i386** [comment]
Stapler passed locally.

**Comment by i386** [comment]
@stephenc @jglick @daniel-beck a review here would be fantastic please :)

File: core/src/main/java/org/kohsuke/stapler/export/ModelBuilder.java [hunk 3]
```diff
@@ -42,12 +42,22 @@
     /*package*/ final Map<Class, Model> models = new ConcurrentHashMap<Class, Model>();
 
     public <T> Model<T> get(Class<T> type) throws NotExportableException {
+        if (type.getAnnotation(ExportedBean.class) == null) {
+            throw new NotExportableException(type);
```
Comment (jglick) [inline core/src/main/java/org/kohsuke/stapler/export/ModelBuilder.java]: Is this not redundant? AFAICT the method it delegates to does the same thing.

File: core/src/main/java/org/kohsuke/stapler/export/ModelBuilder.java [hunk 4]
```diff
@@ -42,12 +42,22 @@
...
+        }
         return get(type, null, null);
     }
 
-    public <T> Model<T> get(Class<T> type, @CheckForNull Class<?> propertyOwner, @Nullable String property) throws NotExportableException {
```
Comment (jglick) [inline core/src/main/java/org/kohsuke/stapler/export/ModelBuilder.java]: Why removing the `throws` clause?

File: core/src/main/java/org/kohsuke/stapler/export/ModelBuilder.java [hunk 5]
```diff
@@ -42,12 +42,22 @@
...
+    public <T> Model<T> get(Class<T> type, @CheckForNull Class<?> propertyOwner, @Nullable String property) {
+        if (type.getAnnotation(ExportedBean.class) == null) {
```
Comment (jglick) [inline core/src/main/java/org/kohsuke/stapler/export/ModelBuilder.java]: I think you rather meant to write

```java
Model<T> model = getOrNull(type, propertyOwner, property);
if (model == null) {
    throw new NotExportableException(type);
}
return model;
```

File: core/src/main/java/org/kohsuke/stapler/export/ModelBuilder.java [hunk 6]
```diff
@@ -42,12 +42,22 @@
...
 
     public <T> Model<T> get(Class<T> type) throws NotExportableException {
+        if (type.getAnnotation(ExportedBean.class) == null) {
+            throw new NotExportableException(type);
+        }
         return get(type, null, null);
     }
 
-    public <T> Model<T> get(Class<T> type, @CheckForNull Class<?> propertyOwner, @Nullable String property) throws NotExportableException {
+    public <T> Model<T> get(Class<T> type, @CheckForNull Class<?> propertyOwner, @Nullable String property) {
```
Comment (jglick) [inline core/src/main/java/org/kohsuke/stapler/export/ModelBuilder.java]: `@Nonnull` return

File: core/src/main/java/org/kohsuke/stapler/export/ModelBuilder.java [hunk 7]
```diff
@@ -42,12 +42,22 @@
...
+        if (type.getAnnotation(ExportedBean.class) == null) {
+            throw new NotExportableException(type);
+        }
+        return getOrNull(type, propertyOwner, property);
+    }
+
+    public <T> Model<T> getOrNull(Class<T> type, @CheckForNull Class<?> propertyOwner, @Nullable String property) throws NotExportableException {
```
Comment (jglick) [inline core/src/main/java/org/kohsuke/stapler/export/ModelBuilder.java]: `@CheckForNull` return. And why is this declared to throw `NotExportableException`?

File: core/src/main/java/org/kohsuke/stapler/export/Property.java [hunk 8]
```diff
@@ -131,23 +131,22 @@ public void writeTo(Object object, TreePruner pruner, DataWriter writer) throws
         TreePruner child = pruner.accept(object, this);
         if (child==null)        return;
 
-        Object d = writer.getExportConfig().getExportInterceptor().getValue(this,object, writer.getExportConfig());
+        Object d = writer.getExportConfig().getExportInterceptor().getValue(this,object,writer.getExportConfig());
```
Comment (jglick) [inline core/src/main/java/org/kohsuke/stapler/export/Property.java]: gratuitous, revert

File: core/src/main/java/org/kohsuke/stapler/export/Property.java [hunk 9]
```diff
@@ -199,118 +198,126 @@ private void writeValue(Type expected, Object value, TreePruner pruner, DataWrit
...
+                Model model = owner.getOrNull(c, parent.type, name);
+                if (model == null) {
+                    throw new NotExportableException(c);
                 }
-                writer.endArray();
+                writer.startObject();
+                model.writeNestedObjectTo(value, pruner, writer);
+                writer.endObject();
+            } catch (NotExportableException ex) {
+                handleNotExportable(expected, value, pruner, writer, skipIfFail, c);
```
Comment (jglick) [inline core/src/main/java/org/kohsuke/stapler/export/Property.java]: [reduced diff](https://github.com/stapler/stapler/pull/127/files?w=1) is easier to read here (generally dislike patches that change indentation of large existing code blocks but sometimes it is hard to avoid)

**Comment by i386** [comment]
@jglick thanks for the review. There was a fair bit of dumb code here on my part, so I did my best to fix it.

File: core/src/main/java/org/kohsuke/stapler/export/Property.java [hunk 10]
```diff
@@ -199,118 +197,122 @@ private void writeValue(Type expected, Object value, TreePruner pruner, DataWrit
...
+            writer.endArray();
+            return;
+        }
+        if(value instanceof Map) {
+            if (verboseMap!=null) {// verbose form
+                writer.startArray();
+                for (Map.Entry e : ((Map<?,?>) value).entrySet()) {
+                    BufferedDataWriter buffer = new BufferedDataWriter(writer.getExportConfig());
+                    try {
+                        writeStartObjectNullType(buffer);
```
Comment (jglick) [inline core/src/main/java/org/kohsuke/stapler/export/Property.java]: Hard to follow this long diff, but IIUC you are basically just extracting `handleNotExportable` and then calling it twice, right? Pretty much unreviewable so I hope the test coverage here is good.

File: core/src/main/java/org/kohsuke/stapler/export/Property.java [hunk 11]
```diff
@@ -199,118 +197,122 @@ private void writeValue(Type expected, Object value, TreePruner pruner, DataWrit
...
         Class c = value.getClass();
 
-        Model model;
-        try {
-            model = owner.get(c, parent.type, name);
-        } catch (NotExportableException ex) {
-            if(STRING_TYPES.contains(c)) {
-                writer.value(value.toString());
-                return;
+        if (c.getAnnotation(ExportedBean.class) == null) {
```
Comment (jglick) [inline core/src/main/java/org/kohsuke/stapler/export/Property.java]: Am I missing something or would this not be caught by the null return value from `getOrNull` a few lines later anyway?

File: core/src/main/java/org/kohsuke/stapler/export/Property.java (hunk 10 above)
Comment (jglick) [inline core/src/main/java/org/kohsuke/stapler/export/Property.java]: Actually the [whitespace-reduced diff](https://github.com/stapler/stapler/pull/127/files?w=1) is more approachable; there must have been a change in indentation for a long block of code.

File: core/src/main/java/org/kohsuke/stapler/export/Property.java [hunk 12]
```diff
@@ -199,118 +197,122 @@ private void writeValue(Type expected, Object value, TreePruner pruner, DataWrit
...
-                } else {
-                    // more generic case
-                    int len = Math.min(r.max, Array.getLength(value));
-                    for (int i=r.min; i<len; i++) {
-                        writeBuffered(act, Array.get(value, i), pruner, writer);
-                    }
+        }
+    }
+
+    private void handleNotExportable(Type expected, Object value, TreePruner pruner, DataWriter writer, boolean skipIfFail, Class c) throws IOException {
```
Comment (jglick) [inline core/src/main/java/org/kohsuke/stapler/export/Property.java]: The method name is potentially misleading. Really this method is handling anything without `@ExportedBean`, but that includes all sorts of objects which are “exportable”. It is only at the very end, starting where we check `skipIfFail`, that this method is in fact handling objects which are not exportable.

File: core/src/main/java/org/kohsuke/stapler/export/Property.java [hunk 13]
```diff
@@ -139,16 +140,12 @@ public void writeTo(Object object, TreePruner pruner, DataWriter writer) throws
...
             if (d != null) {
-                Model model;
-                try {
-                    model = owner.get(d.getClass(), parent.type, name);
-                } catch (NotExportableException e) {
-                    if(writer.getExportConfig().isSkipIfFail()){
-                        return;
-                    }
-                    throw e;
+                Model model = owner.getOrNull(d.getClass(), parent.type, name);
```
[the remark this answers is not in the document]
Comment (i386) [reply core/src/main/java/org/kohsuke/stapler/export/Property.java]: Done

File: core/src/main/java/org/kohsuke/stapler/export/ModelBuilder.java [hunk 14]
```diff
@@ -36,18 +37,29 @@
...
-        if(m==null) {
+        Model<T> model = getOrNull(type, propertyOwner, property);
+        if (model == null) {
+            throw new NotExportableException(type);
+        }
+        return model;
+    }
+
+    @CheckForNull
+    public <T> Model<T> getOrNull(Class<T> type, @CheckForNull Class<?> propertyOwner, @Nullable String property) {
```
[the remark this answers is not in the document]
Comment (i386) [reply core/src/main/java/org/kohsuke/stapler/export/ModelBuilder.java]: Done.

File: core/src/main/java/org/kohsuke/stapler/export/Property.java (hunk 11 above)
Comment (i386) [reply core/src/main/java/org/kohsuke/stapler/export/Property.java]: You are absolutely right. Fixed.

File: core/src/main/java/org/kohsuke/stapler/export/Property.java (hunk 12 above)
Comment (i386) [reply core/src/main/java/org/kohsuke/stapler/export/Property.java]: Renamed and added a description.

File: core/src/main/java/org/kohsuke/stapler/export/Property.java (hunk 10 above)
Comment (i386) [reply core/src/main/java/org/kohsuke/stapler/export/Property.java]: actually, Im going to revert the method extraction. Still too long for my taste but should preserve the diff.

File: core/src/main/java/org/kohsuke/stapler/export/ModelBuilder.java [hunk 15]
```diff
@@ -56,6 +60,12 @@
         return model;
     }
 
+    /**
+     * Instead of throwing {@link NotExportableException} this method will return null
```
Comment (jglick) [inline core/src/main/java/org/kohsuke/stapler/export/ModelBuilder.java]: BTW tip: terminate summary sentences with `.`.
